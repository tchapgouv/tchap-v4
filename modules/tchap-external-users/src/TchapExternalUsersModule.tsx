/*
Copyright 2022 New Vector Ltd. t/a Element

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { RuntimeModule } from "@matrix-org/react-sdk-module-api/lib/RuntimeModule";
import { ModuleApi } from "@matrix-org/react-sdk-module-api/lib/ModuleApi";
import { CantAddDialog  } from "./components/CantAddDialog";
import { AddRoomListener, ExternalUsersLifecycle } from "@matrix-org/react-sdk-module-api/lib/tchap/lifecycles/ExternalUsers";
import { DialogProps } from "@matrix-org/react-sdk-module-api/lib/components/DialogContent";

import React from "react";

export default class TchapExternalUsersModule extends RuntimeModule {
    private domain: String;

    public constructor(moduleApi: ModuleApi) {
        super(moduleApi);

        // const rawConfig = this.moduleApi.getConfigValue(
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_NAMESPACE,
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_KEY,
        //   );
      
        //   assertValidGuestModuleConfig(rawConfig);
        //   this.config = rawConfig;

        this.on(ExternalUsersLifecycle.AddRoom, this.onAddRoom);
    }

    protected onAddRoom: AddRoomListener = async () => {
        await this.moduleApi.openDialog<null, DialogProps, CantAddDialog>("test", (props, ref) => <CantAddDialog ref={ref} {...props} />);
    };
}
