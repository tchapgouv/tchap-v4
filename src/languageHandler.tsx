/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import {
    IVariables,
    TranslatedString,
    TranslationKey as ReactTranslationKey,
    // eslint-disable-next-line camelcase
    _t as react_t,
    // eslint-disable-next-line camelcase
    _td as react_td,
    // eslint-disable-next-line camelcase
    _tDom as react_tDom,
    Tags,
    UserFriendlyError as ReactUserFriendlyError,
    ErrorOptions,
} from "matrix-react-sdk/src/languageHandler";
import { Leaves } from "matrix-react-sdk/src/@types/common";

import type ReactEN from "matrix-react-sdk/src/i18n/strings/en_EN.json";
import type EN from "./i18n/strings/en_EN.json";

/**
 * This module wraps languageHandler in the matrix-react-sdk and adds type casts to include translations
 * which we know will be injected by webpack.
 */

export type TranslationKey = Leaves<typeof EN & typeof ReactEN, "|", string | { other: string }, 4>;

export class UserFriendlyError extends ReactUserFriendlyError {
    public constructor(message: TranslationKey, substitutionVariablesAndCause?: IVariables & ErrorOptions) {
        super(message as ReactTranslationKey, substitutionVariablesAndCause);
    }
}

export function _td(s: TranslationKey): TranslationKey {
    return react_td(s as ReactTranslationKey);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _t(text: TranslationKey, variables?: IVariables): string;
export function _t(text: TranslationKey, variables: IVariables | undefined, tags: Tags): React.ReactNode;
export function _t(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_t(text as ReactTranslationKey, variables, tags!);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _tDom(text: TranslationKey, variables?: IVariables): TranslatedString;
export function _tDom(text: TranslationKey, variables: IVariables, tags: Tags): React.ReactNode;
export function _tDom(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_tDom(text as ReactTranslationKey, variables!, tags!);
}
