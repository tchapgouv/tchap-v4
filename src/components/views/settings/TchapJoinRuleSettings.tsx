/* eslint-disable max-len */
/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { ReactNode } from "react";
import { IJoinRuleEventContent, JoinRule, RestrictedAllowType } from "matrix-js-sdk/src/@types/partials";
import { Room } from "matrix-js-sdk/src/models/room";
import { EventType } from "matrix-js-sdk/src/@types/event";
import StyledRadioGroup, { IDefinition } from "matrix-react-sdk/src/components/views/elements/StyledRadioGroup";
import { _t } from "matrix-react-sdk/src/languageHandler";
import AccessibleButton from "matrix-react-sdk/src/components/views/elements/AccessibleButton";
import SpaceStore from "matrix-react-sdk/src/stores/spaces/SpaceStore";
import RoomAvatar from "matrix-react-sdk/src/components/views/avatars/RoomAvatar";
import { MatrixClientPeg } from "matrix-react-sdk/src/MatrixClientPeg";
import Modal from "matrix-react-sdk/src/Modal";
import ManageRestrictedJoinRuleDialog from "matrix-react-sdk/src/components/views/dialogs/ManageRestrictedJoinRuleDialog";
import RoomUpgradeWarningDialog, {
    IFinishedOpts,
} from "matrix-react-sdk/src/components/views/dialogs/RoomUpgradeWarningDialog";
import { upgradeRoom } from "matrix-react-sdk/src/utils/RoomUpgrade";
import { arrayHasDiff } from "matrix-react-sdk/src/utils/arrays";
import { useLocalEcho } from "matrix-react-sdk/src/hooks/useLocalEcho";
import dis from "matrix-react-sdk/src/dispatcher/dispatcher";
import { ROOM_SECURITY_TAB } from "matrix-react-sdk/src/components/views/dialogs/RoomSettingsDialog";
import { Action } from "matrix-react-sdk/src/dispatcher/actions";
import { ViewRoomPayload } from "matrix-react-sdk/src/dispatcher/payloads/ViewRoomPayload";
import { doesRoomVersionSupport, PreferredRoomVersions } from "matrix-react-sdk/src/utils/PreferredRoomVersions";
import LabelledToggleSwitch from "matrix-react-sdk/src/components/views/elements/LabelledToggleSwitch";
import QuestionDialog from "matrix-react-sdk/src/components/views/dialogs/QuestionDialog";

import TchapUIFeature from "../../../tchap/util/TchapUIFeature";
import { TchapRoomAccessRule, TchapIAccessRuleEventContent, TchapRoomAccessRulesEventId } from "../../../tchap/@types/tchap";

interface IProps {
    room: Room;
    promptUpgrade?: boolean;
    closeSettingsFn(): void;
    onError(error: Error): void;
    beforeChange?(joinRule: JoinRule): Promise<boolean>; // if returns false then aborts the change
    aliasWarning?: ReactNode;
}

const JoinRuleSettings = ({ room, promptUpgrade, aliasWarning, onError, beforeChange, closeSettingsFn }: IProps) => {
    const cli = room.client;

    const roomSupportsRestricted = doesRoomVersionSupport(room.getVersion(), PreferredRoomVersions.RestrictedRooms);
    const preferredRestrictionVersion =
        !roomSupportsRestricted && promptUpgrade ? PreferredRoomVersions.RestrictedRooms : undefined;

    const disabled = !room.currentState.mayClientSendStateEvent(EventType.RoomJoinRules, cli);

    const [content, setContent] = useLocalEcho<IJoinRuleEventContent>(
        () => room.currentState.getStateEvents(EventType.RoomJoinRules, "")?.getContent(),
        (content) => cli.sendStateEvent(room.roomId, EventType.RoomJoinRules, content, ""),
        onError,
    );
    const { join_rule: joinRule = JoinRule.Invite } = content || {};
    const restrictedAllowRoomIds =
        joinRule === JoinRule.Restricted
            ? content.allow?.filter((o) => o.type === RestrictedAllowType.RoomMembership).map((o) => o.room_id)
            : undefined;

    const [contentTchapAccessRule, setTchapAccessRule] = useLocalEcho<TchapIAccessRuleEventContent>(
        () => room.currentState.getStateEvents(TchapRoomAccessRulesEventId, "")?.getContent(),
        (content) => cli.sendStateEvent(room.roomId, TchapRoomAccessRulesEventId, content, ""),
        onError,
    );
    const { rule: accessRule = undefined } = contentTchapAccessRule || {};

    const editRestrictedRoomIds = async (): Promise<string[] | undefined> => {
        let selected = restrictedAllowRoomIds;
        if (!selected?.length && SpaceStore.instance.activeSpaceRoom) {
            selected = [SpaceStore.instance.activeSpaceRoom.roomId];
        }

        const matrixClient = MatrixClientPeg.get();
        const { finished } = Modal.createDialog(
            ManageRestrictedJoinRuleDialog,
            {
                matrixClient,
                room,
                selected,
            },
            "mx_ManageRestrictedJoinRuleDialog_wrapper",
        );

        const [roomIds] = await finished;
        return roomIds;
    };

    /* code from element web
    const definitions: IDefinition<JoinRule>[] = [{
        value: JoinRule.Invite,
        label: _t("Private (invite only)"),
        description: _t("Only invited people can join."),
        checked: joinRule === JoinRule.Invite || (joinRule === JoinRule.Restricted && !restrictedAllowRoomIds?.length),
    }, {
        value: JoinRule.Public,
        label: _t("Public"),
        description: <>
            { _t("Anyone can find and join.") }
            { aliasWarning }
        </>,
    }]; */

    // :TCHAP: we do not permit to change the type of room, thus display only one option
    const definitions: IDefinition<JoinRule>[] = [];

    // :TCHAP: do we need to add the following condition as well (joinRule === JoinRule.Restricted && !restrictedAllowRoomIds?.length)?
    if (joinRule === JoinRule.Invite) {
        let privateRoomDescription = <div>{_t("Only invited people can join.")}</div>;
        // :TCHAP: We could add functions in 'TchapUtils' to determine the type of room and rely on this logic to display components as we did in Android :
        // :TCHAP: https://github.com/tchapgouv/tchap-android/blob/develop/vector/src/main/java/fr/gouv/tchap/core/utils/RoomUtils.kt#L31
        if (accessRule) {
            const openedToExternalUsers = accessRule === TchapRoomAccessRule.Unrestricted;
            const onExternalAccessChange = async () => {
                Modal.createDialog(QuestionDialog, {
                    title: _t("Allow external users to join this room"),
                    description:
                        _t("This action is irreversible.") +
                        " " +
                        _t("Are you sure you want to allow the externals to join this room ?"),
                    button: _t("OK"),
                    onFinished: (confirmed) => {
                        if (!confirmed) return;
                        setTchapAccessRule({ rule: TchapRoomAccessRule.Unrestricted });
                    },
                });
            };
            privateRoomDescription = (
                <div>
                    <div>{_t("Only invited people can join.")}</div>
                    <span>
                        <LabelledToggleSwitch
                            value={openedToExternalUsers}
                            onChange={onExternalAccessChange}
                            label={_t("Allow external users to join this room")}
                            disabled={disabled || openedToExternalUsers}
                        />
                    </span>
                </div>
            );
        }

        definitions.push({
            value: JoinRule.Invite,
            label: _t("Private (invite only)"),
            description: privateRoomDescription,
            checked: true,
        });
    } else if (joinRule === JoinRule.Public) {
        definitions.push({
            value: JoinRule.Public,
            label: _t("Public"),
            description: (
                <>
                    {_t("Anyone can find and join.")}
                    {aliasWarning}
                </>
            ),
        });
    }

    /**
     * TCHAP : disable space-related options if create space feature is not enabled
     */
    if (TchapUIFeature.isSpaceDisplayEnabled) {
        if (roomSupportsRestricted || preferredRestrictionVersion || joinRule === JoinRule.Restricted) {
            let upgradeRequiredPill;
            if (preferredRestrictionVersion) {
                upgradeRequiredPill = (
                    <span className="mx_JoinRuleSettings_upgradeRequired">{_t("Upgrade required")}</span>
                );
            }

            let description;
            if (joinRule === JoinRule.Restricted && restrictedAllowRoomIds?.length) {
                // only show the first 4 spaces we know about, so that the UI doesn't grow out of proportion there are lots.
                const shownSpaces = restrictedAllowRoomIds
                    .map((roomId) => cli.getRoom(roomId))
                    .filter((room) => room?.isSpaceRoom())
                    .slice(0, 4);

                let moreText;
                if (shownSpaces.length < restrictedAllowRoomIds.length) {
                    if (shownSpaces.length > 0) {
                        moreText = _t("& %(count)s more", {
                            count: restrictedAllowRoomIds.length - shownSpaces.length,
                        });
                    } else {
                        moreText = _t("Currently, %(count)s spaces have access", {
                            count: restrictedAllowRoomIds.length,
                        });
                    }
                }

                const onRestrictedRoomIdsChange = (newAllowRoomIds: string[]) => {
                    if (!arrayHasDiff(restrictedAllowRoomIds || [], newAllowRoomIds)) return;

                    if (!newAllowRoomIds.length) {
                        setContent({
                            join_rule: JoinRule.Invite,
                        });
                        return;
                    }

                    setContent({
                        join_rule: JoinRule.Restricted,
                        allow: newAllowRoomIds.map((roomId) => ({
                            type: RestrictedAllowType.RoomMembership,
                            room_id: roomId,
                        })),
                    });
                };

                const onEditRestrictedClick = async () => {
                    const restrictedAllowRoomIds = await editRestrictedRoomIds();
                    if (!Array.isArray(restrictedAllowRoomIds)) return;
                    if (restrictedAllowRoomIds.length > 0) {
                        onRestrictedRoomIdsChange(restrictedAllowRoomIds);
                    } else {
                        onChange(JoinRule.Invite);
                    }
                };

                description = (
                    <div>
                        <span>
                            {_t(
                                "Anyone in a space can find and join. <a>Edit which spaces can access here.</a>",
                                {},
                                {
                                    a: (sub) => (
                                        <AccessibleButton
                                            disabled={disabled}
                                            onClick={onEditRestrictedClick}
                                            kind="link_inline"
                                        >
                                            {sub}
                                        </AccessibleButton>
                                    ),
                                },
                            )}
                        </span>

                        <div className="mx_JoinRuleSettings_spacesWithAccess">
                            <h4>{_t("Spaces with access")}</h4>
                            {shownSpaces.map((room) => {
                                return (
                                    <span key={room.roomId}>
                                        <RoomAvatar room={room} height={32} width={32} />
                                        {room.name}
                                    </span>
                                );
                            })}
                            {moreText && <span>{moreText}</span>}
                        </div>
                    </div>
                );
            } else if (SpaceStore.instance.activeSpaceRoom) {
                description = _t(
                    "Anyone in <spaceName/> can find and join. You can select other spaces too.",
                    {},
                    {
                        spaceName: () => <b>{SpaceStore.instance.activeSpaceRoom.name}</b>,
                    },
                );
            } else {
                description = _t("Anyone in a space can find and join. You can select multiple spaces.");
            }

            definitions.splice(1, 0, {
                value: JoinRule.Restricted,
                label: (
                    <>
                        {_t("Space members")}
                        {upgradeRequiredPill}
                    </>
                ),
                description,
                // if there are 0 allowed spaces then render it as invite only instead
                checked: joinRule === JoinRule.Restricted && !!restrictedAllowRoomIds?.length,
            });
        }
    }

    const onChange = async (joinRule: JoinRule) => {
        const beforeJoinRule = content.join_rule;

        let restrictedAllowRoomIds: string[];
        if (joinRule === JoinRule.Restricted) {
            if (beforeJoinRule === JoinRule.Restricted || roomSupportsRestricted) {
                // Have the user pick which spaces to allow joins from
                restrictedAllowRoomIds = await editRestrictedRoomIds();
                if (!Array.isArray(restrictedAllowRoomIds)) return;
            } else if (preferredRestrictionVersion) {
                // Block this action on a room upgrade otherwise it'd make their room unjoinable
                const targetVersion = preferredRestrictionVersion;

                let warning: JSX.Element;
                const userId = cli.getUserId();
                const unableToUpdateSomeParents = Array.from(SpaceStore.instance.getKnownParents(room.roomId)).some(
                    (roomId) => !cli.getRoom(roomId)?.currentState.maySendStateEvent(EventType.SpaceChild, userId),
                );
                if (unableToUpdateSomeParents) {
                    warning = (
                        <b>
                            {_t(
                                "This room is in some spaces you're not an admin of. " +
                                    "In those spaces, the old room will still be shown, " +
                                    "but people will be prompted to join the new one.",
                            )}
                        </b>
                    );
                }

                Modal.createDialog(RoomUpgradeWarningDialog, {
                    roomId: room.roomId,
                    targetVersion,
                    description: (
                        <>
                            {_t(
                                "This upgrade will allow members of selected spaces " +
                                    "access to this room without an invite.",
                            )}
                            {warning}
                        </>
                    ),
                    doUpgrade: async (
                        opts: IFinishedOpts,
                        fn: (progressText: string, progress: number, total: number) => void,
                    ): Promise<void> => {
                        const roomId = await upgradeRoom(
                            room,
                            targetVersion,
                            opts.invite,
                            true,
                            true,
                            true,
                            (progress) => {
                                const total = 2 + progress.updateSpacesTotal + progress.inviteUsersTotal;
                                if (!progress.roomUpgraded) {
                                    fn(_t("Upgrading room"), 0, total);
                                } else if (!progress.roomSynced) {
                                    fn(_t("Loading new room"), 1, total);
                                } else if (progress.inviteUsersProgress < progress.inviteUsersTotal) {
                                    fn(
                                        _t("Sending invites... (%(progress)s out of %(count)s)", {
                                            progress: progress.inviteUsersProgress,
                                            count: progress.inviteUsersTotal,
                                        }),
                                        2 + progress.inviteUsersProgress,
                                        total,
                                    );
                                } else if (progress.updateSpacesProgress < progress.updateSpacesTotal) {
                                    fn(
                                        _t("Updating spaces... (%(progress)s out of %(count)s)", {
                                            progress: progress.updateSpacesProgress,
                                            count: progress.updateSpacesTotal,
                                        }),
                                        2 + progress.inviteUsersProgress + progress.updateSpacesProgress,
                                        total,
                                    );
                                }
                            },
                        );
                        closeSettingsFn();

                        // switch to the new room in the background
                        dis.dispatch<ViewRoomPayload>({
                            action: Action.ViewRoom,
                            room_id: roomId,
                            metricsTrigger: undefined, // other
                        });

                        // open new settings on this tab
                        dis.dispatch({
                            action: "open_room_settings",
                            initial_tab_id: ROOM_SECURITY_TAB,
                        });
                    },
                });

                return;
            }

            // when setting to 0 allowed rooms/spaces set to invite only instead as per the note
            if (!restrictedAllowRoomIds.length) {
                joinRule = JoinRule.Invite;
            }
        }

        if (beforeJoinRule === joinRule && !restrictedAllowRoomIds) return;
        if (beforeChange && !(await beforeChange(joinRule))) return;

        const newContent: IJoinRuleEventContent = {
            join_rule: joinRule,
        };

        // pre-set the accepted spaces with the currently viewed one as per the microcopy
        if (joinRule === JoinRule.Restricted) {
            newContent.allow = restrictedAllowRoomIds.map((roomId) => ({
                type: RestrictedAllowType.RoomMembership,
                room_id: roomId,
            }));
        }

        setContent(newContent);
    };

    return (
        <StyledRadioGroup
            name="joinRule"
            value={joinRule}
            onChange={onChange}
            definitions={definitions}
            disabled={disabled}
            className="mx_JoinRuleSettings_radioButton"
        />
    );
};

export default JoinRuleSettings;
