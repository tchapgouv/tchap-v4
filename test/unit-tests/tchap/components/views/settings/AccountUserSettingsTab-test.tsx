import { render } from "jest-matrix-react";
import React from "react";
import { MatrixClient } from "matrix-js-sdk/src/matrix";
import { logger } from "matrix-js-sdk/src/logger";
import { MockedObject } from "jest-mock";

import AccountUserSettingsTab from "~tchap-web/src/components/views/settings/tabs/user/AccountUserSettingsTab";
import { SdkContextClass, SDKContext } from "~tchap-web/src/contexts/SDKContext";
import SettingsStore from "~tchap-web/src/settings/SettingsStore";
import { UIFeature } from "~tchap-web/src/settings/UIFeature";
import { OidcClientStore } from "~tchap-web/src/stores/oidc/OidcClientStore";
import MatrixClientContext from "~tchap-web/src/contexts/MatrixClientContext";
import {
    getMockClientWithEventEmitter,
    mockClientMethodsServer,
    mockClientMethodsUser,
    mockPlatformPeg,
} from "~tchap-web/test/test-utils";

jest.mock(
    "~tchap-web/src/components/views/settings/ChangePassword",
    () =>
        ({ onError, onFinished }: { onError: (e: Error) => void; onFinished: () => void }) => {
            return <button>Mock change password</button>;
        },
);

describe("<AccountUserSettingsTab />", () => {
    const defaultProps = {
        closeSettingsFn: jest.fn(),
    };

    const userId = "@alice:server.org";
    let mockClient: MockedObject<MatrixClient>;

    let stores: SdkContextClass;

    const getComponent = () => (
        <MatrixClientContext.Provider value={mockClient}>
            <SDKContext.Provider value={stores}>
                <AccountUserSettingsTab {...defaultProps} />
            </SDKContext.Provider>
        </MatrixClientContext.Provider>
    );

    beforeEach(() => {
        jest.spyOn(SettingsStore, "getValue").mockReturnValue(false);
        mockPlatformPeg();
        jest.clearAllMocks();
        jest.spyOn(SettingsStore, "getValue").mockRestore();
        jest.spyOn(logger, "error").mockRestore();

        mockClient = getMockClientWithEventEmitter({
            ...mockClientMethodsUser(userId),
            ...mockClientMethodsServer(),
            getCapabilities: jest.fn(),
            getThreePids: jest.fn(),
            getIdentityServerUrl: jest.fn(),
            deleteThreePid: jest.fn(),
        });

        mockClient.getCapabilities.mockResolvedValue({});
        mockClient.getThreePids.mockResolvedValue({
            threepids: [],
        });
        mockClient.deleteThreePid.mockResolvedValue({
            id_server_unbind_result: "success",
        });

        stores = new SdkContextClass();
        stores.client = mockClient;
        // stub out this store completely to avoid mocking initialisation
        const mockOidcClientStore = {} as unknown as OidcClientStore;
        jest.spyOn(stores, "oidcClientStore", "get").mockReturnValue(mockOidcClientStore);
        jest.spyOn(SettingsStore, "getValue").mockImplementation(
            (settingName: string) => settingName === UIFeature.Deactivate,
        );
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });

    describe("common view snapshot", () => {
        it("should render section when account deactivation feature is enabled", () => {
            const { container } = render(getComponent());

            expect(container).toMatchSnapshot();
        });
    });
});
