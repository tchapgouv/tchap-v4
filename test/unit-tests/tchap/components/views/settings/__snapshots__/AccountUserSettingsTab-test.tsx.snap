// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccountUserSettingsTab /> common view snapshot should render section when account deactivation feature is enabled 1`] = `
<div>
  <div
    class="mx_SettingsTab"
    data-testid="mx_AccountUserSettingsTab"
  >
    <div
      class="mx_SettingsTab_sections"
    >
      <div
        class="mx_SettingsSection"
      >
        <div
          class="mx_SettingsSection_subSections"
        >
          <div
            class="mx_UserProfileSettings"
          >
            <h2>
              Profile
            </h2>
            <div>
              Your account is managed separately by an identity provider and so some of your personal information can’t be changed here.
            </div>
            <div
              class="mx_UserProfileSettings_profile"
            >
              <div
                aria-label="Profile picture"
                class="mx_AvatarSetting_avatar"
                role="group"
              >
                <div
                  aria-disabled="true"
                  class="mx_AccessibleButton mx_AvatarSetting_avatarPlaceholder mx_AvatarSetting_avatarDisplay mx_AccessibleButton_disabled"
                  disabled=""
                  role="button"
                  tabindex="-1"
                >
                  <span
                    class="_avatar_mcap2_17 mx_BaseAvatar _avatar-imageless_mcap2_61"
                    data-color="2"
                    data-testid="avatar-img"
                    data-type="round"
                    role="presentation"
                    style="--cpd-avatar-size: 90px;"
                  />
                </div>
              </div>
              <form
                class="_root_dgy0u_24 mx_UserProfileSettings_profile_displayName"
              >
                <div
                  class="_field_dgy0u_34"
                >
                  <label
                    class="_label_dgy0u_67"
                    for="radix-:r0:"
                  >
                    Display Name
                  </label>
                  <div
                    class="_controls_1h4nb_17"
                  >
                    <input
                      class="_control_9gon8_18"
                      disabled=""
                      id="radix-:r0:"
                      name="input"
                      title=""
                      value=""
                    />
                  </div>
                </div>
              </form>
            </div>
            <div
              class="mx_UserProfileSettings_profile_controls_userId"
            >
              <div
                class="mx_UserProfileSettings_profile_controls_userId_label"
                id="1234"
              >
                Username
              </div>
              <div
                aria-labelledby="1234"
                class="mx_CopyableText mx_CopyableText_border"
              >
                @alice:server.org
                <div
                  aria-label="Copy"
                  class="mx_AccessibleButton mx_CopyableText_copyButton"
                  role="button"
                  tabindex="0"
                />
              </div>
            </div>
            <div
              class="mx_Flex mx_UserProfileSettings_profile_buttons"
              style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: start; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-4x);"
            >
              <div
                class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_danger_outline"
                role="button"
                tabindex="0"
              >
                <svg
                  class="mx_UserProfileSettings_accountmanageIcon"
                  fill="currentColor"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 12.031c0-.283.096-.52.288-.712A.968.968 0 0 1 10 11.03h7.15l-1.875-1.875a.96.96 0 0 1-.3-.7c0-.266.108-.508.325-.725a.93.93 0 0 1 .712-.287.977.977 0 0 1 .688.287l3.6 3.6c.1.1.17.209.212.325.042.117.063.242.063.375 0 .134-.02.259-.063.375a.877.877 0 0 1-.212.325l-3.6 3.6a.93.93 0 0 1-.712.288.977.977 0 0 1-.688-.288 1.02 1.02 0 0 1-.313-.712.931.931 0 0 1 .288-.713l1.875-1.875H10a.968.968 0 0 1-.712-.287A.968.968 0 0 1 9 12.03Zm-6-7c0-.55.196-1.02.588-1.412A1.926 1.926 0 0 1 5 3.03h6a.97.97 0 0 1 .713.288.968.968 0 0 1 .287.712.97.97 0 0 1-.287.713.968.968 0 0 1-.713.287H5v14h6a.97.97 0 0 1 .713.288.968.968 0 0 1 .287.712.97.97 0 0 1-.287.713.968.968 0 0 1-.713.287H5c-.55 0-1.02-.196-1.412-.587A1.926 1.926 0 0 1 3 19.03v-14Z"
                  />
                </svg>
                Sign out
              </div>
            </div>
          </div>
          <div
            class="mx_SettingsSubsection"
          >
            <div
              class="mx_SettingsSubsectionHeading"
            >
              <h3
                class="mx_Heading_h4 mx_SettingsSubsectionHeading_heading"
              >
                settings
              </h3>
            </div>
            <div
              class="mx_SettingsSubsection_content mx_SettingsSubsection_contentStretch"
            >
              <div
                class="mx_SettingsFlag"
              >
                <label
                  class="mx_SettingsFlag_label"
                  for="redList-toggle"
                >
                  <span
                    class="mx_SettingsFlag_labelText"
                  >
                    settings
                  </span>
                  <div
                    class="mx_SettingsFlag_microcopy"
                  >
                    settings
                  </div>
                </label>
                <div
                  aria-checked="false"
                  aria-disabled="false"
                  aria-label="settings"
                  class="mx_AccessibleButton mx_ToggleSwitch mx_ToggleSwitch_enabled"
                  id="redList-toggle"
                  role="switch"
                  tabindex="0"
                >
                  <div
                    class="mx_ToggleSwitch_ball"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="mx_SettingsSection"
      >
        <h2
          class="mx_Heading_h3"
        >
          Deactivate Account
        </h2>
        <div
          class="mx_SettingsSection_subSections"
        >
          <div
            class="mx_SettingsSubsection"
            data-testid="account-management-section"
          >
            <div
              class="mx_SettingsSubsectionHeading"
            >
              <h3
                class="mx_Heading_h4 mx_SettingsSubsectionHeading_heading"
              >
                Account management
              </h3>
            </div>
            <div
              class="mx_SettingsSubsection_description"
            >
              <div
                class="mx_SettingsSubsection_text"
              >
                Deactivating your account is a permanent action — be careful!
              </div>
            </div>
            <div
              class="mx_SettingsSubsection_content"
            >
              <div
                class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_danger"
                role="button"
                tabindex="0"
              >
                Deactivate Account
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
